jestjs/jest
# fix(jest-circus) correct concurrent event ordering

### Summary

Note this is a re-raise of an [old PR](https://github.com/jestjs/jest/pull/14643) that I previously let go stale.

I'm unsure if this is a major, minor, or patch change. It does not change the API, but it changes the order of events emitted from `jest-circus`. Please advise.

This fixes the issue documented [here](https://github.com/jestjs/jest/issues/14616)

When using `test.concurrent` events such as `test_start`, `test_started`, and `test_fn_start` are dispatched after the body of the test is executed. This is due to the `_runTest` function not being executed as part of the parallel execution runner; instead, it was executed synchronously like non-concurrent tests.

The solution was to use _runTest within startTestsConcurrently() and then wait for that process to complete instead of replacing the `test.fn` as was being done before.

I also added a bunch of e2e tests for concurrent, this change most likely closes some [other issues](https://github.com/jestjs/jest/issues/7997) relating to `before/afterAll`.

### Test plan

- Checkout commit `953bbed700`
- Run `yarn test packages/jest-circus/src/__tests__/baseTest.test.ts` to observe the broken snapshot content for the concurrent test.
```
hello one <-- test running before `test_start` has fired.
hello two
hello three
run_describe_start: describe
test_start: one
test_started: one
test_fn_start: one
test_fn_failure: one
```
- Check out `HEAD` and rerun the test. Note that the output has been reordered to match the expectation that the circus actions occur before and after the test execution.
- Execute the new concurrent e2e tests with `yarn jest e2e/__tests__/circusConcurrent.test.ts` and `yarn jest e2e/__tests__/testRetries.test.ts`

**Note: at the time this is being raised `main` does not pass `yarn test`, I have confirmed that the same tests are failing before and after my change.**
